src/storage/store/string_column_chunk.cpp

```diff
@@ -128,12 +136,72 @@ void StringColumnChunk::setValueFromString(const char* value, uint64_t length, u
     offsetChunk->setValue<string_offset_t>(startOffset, index);
     offsetChunk->setNumValues(index + 1);
     ColumnChunk::setValue<string_index_t>(index, pos);
+    indexTable.insert(std::make_pair(std::string(value, length), index));
+}
+
+void StringColumnChunk::prepareForFlush() {
+    // Only de-duplicate strings, not blobs
+    if (dataType.getLogicalTypeID() != LogicalTypeID::STRING) {
+        return;
+    }
+    // Prune unused entries in the dictionary before we flush
+    // We already de-duplicate as we go, but when out of place updates occur new values will be
+    // appended to the end and the original values may be able to be pruned before flushing them to
+    // disk
+    std::unordered_map<std::string_view, string_index_t> indexTable;
+    // Create new buffers with only the unique strings in the index map, updating them as values
+    // are added to the indexMap
+    auto newDataChunk =
+        ColumnChunkFactory::createColumnChunk(LogicalType(LogicalTypeID::UINT8), false);
+
+    auto newOffsetChunk = ColumnChunkFactory::createColumnChunk(LogicalType(LogicalTypeID::UINT64),
+        enableCompression, StorageConstants::NODE_GROUP_SIZE * 0.75);
+
+    // We re-write the source buffer as we go over it.
+    // Each index is replaced by a new one for the de-duplicated data in the new data buffer
+    auto outIndices = (string_index_t*)buffer.get();
+    for (int i = 0; i < numValues; i++) {
+        if (nullChunk->isNull(i)) {
+            continue;
+        }
+        auto stringData = getValue<std::string_view>(i);
+        auto index = ColumnChunk::getValue<string_index_t>(i);
+        auto offset = offsetChunk->getValue<string_offset_t>(index);
+
+        auto result = indexTable.find(stringData);
+        if (enableCompression && result != indexTable.end()) {
+            outIndices[i] = result->second;
+        } else {
+            if (enableCompression) {
+                indexTable.insert(std::make_pair(stringData, indexTable.size()));
+            }
+
+            auto newIndex = newOffsetChunk->getNumValues();
+            if (newIndex >= newOffsetChunk->getCapacity()) {
```

Comment:
Is there an implicit assumption that `newIndex` will always sit between `capacity` and `capacity * 2` ?

Reply:
Yes, but if the number of values is strictly greater than the capacity then we have other issues. In normal circumstances this will only be true when the newIndex is equal to the capacity. Values are being added one at a time here.